### PR TITLE
LibGit2: respect NetworkOptions.verify_host

### PR DESCRIPTION
**Pull request by StefanKarpinski**

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
Everything passed except aarch64 which has some numerical test failures that definitely aren't related. I'm gonna do some more testing before merge to make sure this is doing what I think it is, but at least it doesn't break anything.

File: stdlib/LibGit2/src/callbacks.jl [hunk 1]
```diff
@@ -359,9 +359,29 @@ function fetchhead_foreach_callback(ref_name::Cstring, remote_url::Cstring,
     return Cint(0)
 end
 
+# from `enum git_cert_t` in `cert.h`
+const CERT_TYPE_TLS = 1 # GIT_CERT_X509
+const CERT_TYPE_SSH = 2 # GIT_CERT_HOSTKEY_LIBSSH2
```
Comment (omus) [inline stdlib/LibGit2/src/callbacks.jl:364]: Should be moved to consts.jl

File: stdlib/LibGit2/src/callbacks.jl [hunk 2]
```diff
@@ -359,9 +359,29 @@ function fetchhead_foreach_callback(ref_name::Cstring, remote_url::Cstring,
...
+    cert_p :: Ptr{Cvoid},
+    valid  :: Cint,
+    host_p :: Ptr{Cchar},
+    user_p :: Ptr{Cvoid},
+)::Cint
+    host = unsafe_string(host_p)
+    cert_type = unsafe_load(convert(Ptr{Cint}, cert_p))
+    transport = cert_type == CERT_TYPE_TLS ? "TLS" :
+                cert_type == CERT_TYPE_SSH ? "SSH" : nothing
+    @show host, transport
```
Comment (omus) [inline stdlib/LibGit2/src/callbacks.jl:376]: Don't forget to remove this

**Review by omus — APPROVED** [review]

File: stdlib/LibGit2/src/callbacks.jl (hunk 1 above)
Comment (StefanKarpinski) [reply stdlib/LibGit2/src/callbacks.jl:364]: Will do.

File: stdlib/LibGit2/src/callbacks.jl (hunk 2 above)
Comment (StefanKarpinski) [reply stdlib/LibGit2/src/callbacks.jl:376]: Will do. I'm still testing here. Should have marked the PR as draft for now.

**Comment by StefanKarpinski** [comment]
Ok, so here's a way to test this using ghostunnel. First, set up the "client" side that connects to `github.com`:
```
ghostunnel client --target=github.com:443 --listen=localhost:1234 --disable-authentication
```
Next, generate a phony key/cert pair:
```
openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -days 365 -nodes
```
Then start the "server" side that serves incoming HTTPS connections and sends them through to the client:
```
sudo ghostunnel server --listen $ip:443 --target localhost:1234 --key key.pem --cert cert.pem --disable-authentication
```
Then you modify `/etc/hosts` on the machine you're running `julia` on to add this entry:
```
$ip github.com
```
With this configuration, the local machine will be man-in-the-middled for HTTPS connections to `github.com`. In order to test that this is correctly rejected by Pkg, you'll want to `export JULIA_PKG_SERVER=""` and delete `~/.julia/registries/General` so that Pkg uses LibGit2 to clone the registry. Pkg should abort since it's getting man-in-the-middled. However, if you set `JULIA_NO_VERIFY_HOSTS="github.com"` then you should be able to clone registry again, despite the man-in-the-middle attack.

This is useful because this kind of man-in-the-middle attack is exactly what is done behind many firewalls. If this kind of man-in-the-middle proxy is correctly setup, it is still secure (aside from the operator being able to snoop on your traffic) since the proxy will verify it's outbound connections.

**Comment by GunnarFarneback** [comment]
This is not working right with ssh keys, as discussed on slack. To reproduce, try to add your own fork of General with `git` URL.
```
(@v1.6) pkg> registry add git@github.com:GunnarFarneback/General.git
     Cloning registry from "git@github.com:GunnarFarneback/General.git"
ERROR: failed to clone from git@github.com:GunnarFarneback/General.git, error: GitError(Code:ERROR, Class:Net, user cancelled hostkey check)
```

**Comment by StefanKarpinski** [comment]
SSH issue should be fixed by https://github.com/JuliaLang/julia/pull/38563.

**Comment by GunnarFarneback** [comment]
Confirmed, works well now.

**Pull request merged**